### ISSUE TITLE
Concurrent convert_input_data

### ISSUE THREAD

**Issue by idanh** [facebookresearch/PyTorch-BigGraph#108]
Hi,

I have a 10GB dataset in csv I'd like to convert to h5 using convert_input_data method.
Seems to me (from looking at the code) that this is only possible using one processor.

Before looking into PR'ing my own solution, I'd like to know:
- is it possible to run convert_input_data only multiple processors today?
- if not implemented today, is it feasible to do so? Do you see any obvious issues in parallelising this process?

One simple option could be calling convert_input_data multiple times with different lists from different processes, but I have a feeling that I'll need to merge more files (dictionary.json) and fix entity_count to merge back to 32 partitions. So I believe this solution won't work out of the box.

I'll note that I have 32 partitions for one relation specified in configuration file.
Much appreciated.

**Comment by lerks** [comment]
I think this was somewhat addressed in #42. Does the information there answer your question or is there more you'd like to know?

**Comment by idanh** [comment]
@lerks sorry for delayed response just got back from vacation.

Yes, #42 does help. So basically a high level overview of whats needed is:
For each relation and for each partition, partitioned (h5) files are created and than for each of those tuples (relation and partition) we create a distinct count of (lhs, rhs, relation) triplets as entity_count_(relation)_(parittion).txt.

After this is done, we create a dictionary that contains all lhs from all relations.

Is that correct?
If so, this sounds to me that it could be done with spark in multiple stages.

Thank you very much!

**Comment by lerks** [comment]
I think you got confused, some of the times you said "relation" you probably meant "entity".

Anyways, to recap here is what is needed:
- Suppose you have a list of edges, each containing the type and id of the entity on the left-hand side, the type and id of the entity on the right-hand side, and the type of the relation.
- For each type of entities, you need to map all their ids to (partition, offset) pairs, where the partition of each entity is chosen uniformly at random, and the offsets for each partition are distinct and consecutive. Then you need to store the number of entities in each partition in a .txt file, as you said.
- Then you transform each edge from the list above, in the following way:
  - You replace the relation type by the index at which that type appears in the PBG config
  - You replace the entity type and id, for each side, by the (partition, offset) pair using the maps you just produced. (This means you'll "forget" the entity type, but that's fine: it can be recovered from the relation type).
- Once you have this transformed edge list, you need to split it into buckets, i.e.:
  - Group it by (lhs partition, rhs partition)
  - Shuffle each such group randomly
  - Store each group in a separate .h5 file.

Hope this clarifies.

**Comment by idanh** [comment]
@lerks thank you again,

> offsets for each partition are distinct and consecutive.

distinct globally right? (not per relation)
So for entity A in partition 1 and entity B in partition 2 should have the consecutive offsets of 0, 1.
For an example, the offsets used in "convert_input_data" are from the dictionary.json file, is that accurate?


> Then you need to store the number of entities in each partition in a .txt file, as you said.

Looking at the output, the text file name is: entity_count_(relation)_(partition_number).txt
And it counts the number of entities for a specific relation and partition right? In the case of two relations and 1 partition, I will distinctly count (entity, relation) tuples for a particular partition?

> Store each group in a separate .h5 file.

Can you please provide the file format the training code is expecting? I believe it was "edges_(bucket)_(partition).h5" just want to make sure I don't mix up the two.

**Comment by lerks** [comment]
The offsets have to be distinct within each (entity type, partition) group. Relation types have nothing to do with this.

The fiel names are `entity_count_{entity}_{part}.txt` and `edges_{lhs_part}_{rhs_part}.h5`. Also, after a recent change we're not using `dictionary.json` anymore, we now split it up into multiple files, named `entity_names_{entity}_{part}.json` and `dynamic_rel_names.json`. However, you don't have to produce these files manually: you can use the new classes inside `graph_storages.py` to take care of actually writing the data to disk.

**Comment by idanh** [comment]
@lerks Thank you for the update.

Glancing at the new code, I believe I should take a closer look at FileEdgeStorage.

Thank you again for your assistant. It is much appreciated.
I'll close this issue.

**Comment by lerks** [comment]
I feel a more concrete example might help here. Consider the following config:
```
"entities": {
    "users": {"num_partitions": 3},
    "albums": {"num_partitions": 2},
    "songs": {"num_partitions": 2}
},
"relations": [
    {"name": "bought", "lhs": "users", "rhs": "albums"},
    {"name": "listened", "lhs": "users", "rhs": "songs"}
]
```
so we have entities of three types (users, albums and songs) and two relation types: users having bought albums (this is index 0, as it's the first one listed in the `relations` param) and users having listened to songs (index 1).

Imagine the following edgelist:
- user foo listened to song A
- user bar bought album alpha
- user bar listened to song A
- user foo bought album beta
- user foo bought album gamma
- user baz bought album beta
- user eggs bought album gamma
- user eggs listened to song B
- user spam listened to song B
- user spam listened to song A
- user ham bought album alpha
- user ham listened to song B

So in total there are 6 users: foo, bar, baz, spam, eggs, ham. We need to split them into 3 partitions, so we first assign a partition 0, 1 or 2 to each user at random, for example: foo -> 2, bar -> 1, baz -> 2, spam -> 0, eggs -> 2, ham -> 1. Then, within each partition, we arbitrarily sort the users that belong to that partition (can be any order, not necessarily random), for example: partition 0 -> [spam], partition 1 -> [ham, bar], partition 2 -> [baz, foo, eggs]. Now, using the position each user has in its partition's list, we can determine an offset:
- user foo: partition 2, offset 1
- user bar: partition 1, offset 1
- user baz: partition 2, offset 0
- user spam: partition 0, offset 0
- user eggs: partition 2, offset 2
- user ham: partition 1, offset 1

In a similar way we can partition albums and songs and obtain, for example:
- songs A and B; split in: partition 0 -> [A], partition 1 -> [B]
- albums alpha, beta and gamma; split in partition 0 -> [gamma, alpha], partition 1 -> [beta]
Thus:
- song A: partition 0, offset 0
- song B: partition 1, offset 0
- album alpha: partition 0, offset 1
- album beta: partition 1, offset 0
- album gamma: partition 0, offset 0

Thus the initial edgelist can be converted to:
- partition 2, offset 1, relation type 1, partition 0, offset 0 (a)
- partition 1, offset 1, relation type 0, partition 0, offset 1 (b)
- partition 1, offset 1, relation type 1, partition 0, offset 0 (c)
- partition 2, offset 1, relation type 0, partition 1, offset 0 (d)
- partition 2, offset 1, relation type 0, partition 0, offset 0 (e)
- partition 2, offset 0, relation type 0, partition 1, offset 0 (f)
- partition 2, offset 2, relation type 0, partition 0, offset 0 (g)
- partition 2, offset 2, relation type 1, partition 1, offset 0 (h)
- partition 0, offset 0, relation type 1, partition 1, offset 0 (i)
- partition 0, offset 0, relation type 1, partition 0, offset 0 (j)
- partition 1, offset 1, relation type 0, partition 0, offset 1 (k)
- partition 1, offset 1, relation type 1, partition 1, offset 0 (l)

 by replacing each user, album and song with the (partition, offset) pair we computed above, and each relation type with the index we set at the beginning. (I denoted each edge by a letter to refer to them later)

Once you have all this, you can produce the required files:
- `entity_count_users_0.txt`, containing 1 (because it contains just spam)
- `entity_count_users_1.txt`, containing 2 (because it contains ham and bar)
- `entity_count_users_2.txt`, containing 3
- `entity_count_songs_0.txt`, containing 1
- `entity_count_songs_1.txt`, containing 1
- `entity_count_album_0.txt`, containing 2
- `entity_count_album_1.txt`, containing 1
- `edges_0_0.h5`, containing:
  - offset 0, relation type 1, offset 0 (j)
- `edges_0_1.h5`, containing:
  - offset 0, relation type 1, offset 0 (i)
- `edges_1_0.h5`, containing:
  - offset 1, relation type 0, offset 1 (b)
  - offset 1, relation type 1, offset 0 (c)
  - offset 1, relation type 0, offset 1 (k)
- `edges_1_1.h5`, containing:
  - offset 1, relation type 1, offset 0 (l)
- `edges_2_0.h5`, containing:
  - offset 1, relation type 0, offset 0 (e)
  - offset 2, relation type 0, offset 0 (g)
  - offset 1, relation type 1, offset 0 (a)
- `edges_2_1.h5`, containing:
  - offset 1, relation type 0, offset 0 (d)
  - offset 0, relation type 0, offset 0 (f)
  - offset 2, relation type 1, offset 0 (h)

**Comment by idanh** [comment]
@lerks Your example is very, very helpful! Thank you for explaining the process.
Just to make sure, as you don't support edge weights at the moment, I read in #55 that one suggestion (tho, is not deterministic depending on how many of the same edge falls into the same partition), but given that you support adding the same edge twice, in above example, say that:

> user foo listened to song A two times

So looking at edges_2_0.h5 (lhs foo, partition 2, rhs song A, partition 0) we should see:

> offset 1 (foo is offset 1), relation type 1 (listen to song), offset 0 (song A) twice.

Is that correct?

**Comment by lerks** [comment]
Yes, correct: each `edges_{lhs}_{rhs}.h5` file can contain multiple times each row, which means there will be multiple edges between certain pairs of entities, which is equivalent to these edges being weighted more.
